Install latest patch version if major.minor semantic version provided; Add latest command to install

ChronosMasterOfAllTime opened the pull request:
Should resolve the ask in #121.

Adds:

`goenv install latest`
`goenv install unstable`

if the semantic version is only supplied with major/minor, find the latest patch version and use that for install.

e.g. if `1.19` is supplied and the following versions are available:

```
1.19.0
1.19.1
1.19.2
```

`1.19.2` will be installed

Added unit tests to validate functionality

detro commented on COMMANDS.md:
```diff
@@ -1,35 +1,35 @@
-* [`goenv versions`](#goenv-versions)
-* [`goenv whence`](#goenv-whence)
-* [`goenv which`](#goenv-which)
+- [`goenv commands`](#goenv-commands)
+- [`goenv completions`](#goenv-completions)
+- [`goenv exec`](#goenv-exec)
+- [`goenv global`](#goenv-global)
+- [`goenv help`](#goenv-help)
+- [`goenv hooks`](#goenv-hooks)
+- [`goenv init`](#goenv-init)
+- [`goenv install`](#goenv-install)
+- [`goenv local`](#goenv-local)
+- [`goenv prefix`](#goenv-prefix)
+- [`goenv rehash`](#goenv-rehash)
+- [`goenv root`](#goenv-root)
+- [`goenv shell`](#goenv-shell)
+- [`goenv shims`](#goenv-shims)
+- [`goenv uninstall`](#goenv-uninstall)
+- [`goenv version`](#goenv-version)
+- [`goenv --version`](#goenv---version)
+- [`goenv version-file`](#goenv-version-file)
+- [`goenv version-file-read`](#goenv-version-file-read)
+- [`goenv version-file-write`](#goenv-version-file-write)
+- [`goenv version-name`](#goenv-version-name)
+- [`goenv version-origin`](#goenv-version-origin)
+- [`goenv versions`](#goenv-versions)
+- [`goenv whence`](#goenv-whence)
+- [`goenv which`](#goenv-which)
```
2cents @ChronosMasterOfAllTime - it makes for harder to review PRs, if in each there are changes that are unrelated from the focus of the PR.

I totally understand how tempting it is, but I'd try to avoid it, especially if it's stuff like "should I create a list in MD with `*` or `-`", that is totally inconsequential at rendering.

ChronosMasterOfAllTime replied:
Yeah sorry about that. I had my formatters turned on when saving so it automatically removes whitespace etc...will fix the MD changes

ChronosMasterOfAllTime replied:
Fixed

detro commented on plugins/go-build/bin/go-build:
```diff
@@ -760,7 +760,13 @@ unset IPV6
 GO_BUILD_INSTALL_PREFIX="$(abs_dirname "$0")/.."
 
-IFS=: GO_BUILD_DEFINITIONS=($GO_BUILD_DEFINITIONS ${GO_BUILD_ROOT:-$GO_BUILD_INSTALL_PREFIX}/share/go-build)
+if [[ $USE_FAKE_DEFINITIONS == "true" ]]; then
+  DIR_SUFFIX=test/fixtures/definitions
+else
+  DIR_SUFFIX=share/go-build
+fi
```
Can you add a comment on top of this to explain what does "using fake definitions" do? From what I gather, it's useful for testing, to "mock" the available go definitions.

ChronosMasterOfAllTime replied:
Done

detro commented on plugins/go-build/bin/goenv-install:
```diff
@@ -62,6 +62,10 @@ definitions() {
 }
 
+latest_version() {
+  definitions | grep -oE "^$1\\.?([0-9]+|beta[0-9]+|rc[0-9]+)?" | tail -1
```
Just double checking, as I'm not sure I understand the regexp (my bad): are we allowing as "latest" also _beta_ and _rc_ versions?

If so, I'd advice against it: I think it's a much less risky default to assume that "latest" means "latest stable". Who uses _beta_ or _rc_ versions is not likely to be a normal Go developer, trying to go about their business.

ChronosMasterOfAllTime replied:
Your assumption is correct; we can remove RC and Beta and add an "unstable" feature

ChronosMasterOfAllTime replied:
Added `unstable` feature and latest only does latest stable

detro commented on plugins/go-build/bin/goenv-install:
```diff
@@ -123,9 +127,15 @@ DEFINITION="${ARGUMENTS[0]}"
-if grep -q -E "^[0-9]+\.[0-9]+(\s*)$" <<< ${DEFINITION}; then
-  echo "Adding patch version 0 to ${DEFINITION}"
-  DEFINITION="${DEFINITION}.0"
+if [[ ${DEFINITION} == "latest" ]]; then
+  LATEST=$(latest_version "[0-9]\\.[0-9]+")
+  echo "Installing latest version ${LATEST}..."
+  DEFINITION=$LATEST
+elif grep -q -E "^[0-9]+\.[0-9]+(\s*)$" <<<${DEFINITION}; then
+  REGEX=$(echo $DEFINITION | sed s/\\./\\\\./)
+  LATEST_PATCH=$(latest_version $REGEX)
+  echo "Using latest patch version $LATEST_PATCH"
+  DEFINITION="$LATEST_PATCH"
```
I think the `NOTE` above needs to be updated to reflect the more complex scenarios covered here.

ChronosMasterOfAllTime replied:
Not sure I understand the ask on this one

detro reviewed (commented):
This is looking already pretty good: I just left a feed comments that have more to do with long-term maintenance than anything else.

Plus, if I got the regexp right, I think you are intending to match `"latest"` version to also `"beta"` and `"rc"` releases, and I don't think it's a great idea. But I'm not the owner of the repo, so I'd leave it to @syndbg and other maintainers to make the call.

Personally, I think the vast majority of users want a tool that lets them easily switch version of `go`, while they move across their repositories/projects. Those people are rarely interested in _non-stable_ versions of Go.

ChronosMasterOfAllTime commented:
> This is looking already pretty good: I just left a feed comments that have more to do with long-term maintenance than anything else.
> 
> Plus, if I got the regexp right, I think you are intending to match `"latest"` version to also `"beta"` and `"rc"` releases, and I don't think it's a great idea. But I'm not the owner of the repo, so I'd leave it to @syndbg and other maintainers to make the call.
> 
> Personally, I think the vast majority of users want a tool that lets them easily switch version of `go`, while they move across their repositories/projects. Those people are rarely interested in _non-stable_ versions of Go.

Addressed all comments @detro

EfChouTR commented:
Awesome!

detro commented on plugins/go-build/bin/go-build:
```diff
@@ -760,7 +760,14 @@ unset IPV6
 GO_BUILD_INSTALL_PREFIX="$(abs_dirname "$0")/.."
 
-IFS=: GO_BUILD_DEFINITIONS=($GO_BUILD_DEFINITIONS ${GO_BUILD_ROOT:-$GO_BUILD_INSTALL_PREFIX}/share/go-build)
+# USE_FAKE_DEFINITIONS is used for tests to mock the definitions list to our available test fixtures
```
❤️

detro commented on plugins/go-build/bin/goenv-install:
```diff
@@ -2,7 +2,7 @@
 # Summary: Install a Go version using go-build
 #
-# Usage: goenv install [-f] [-kvp] <version>
+# Usage: goenv install [-f] [-kvp] <version>|latest|unstable
```
Oh, I think this is a great approach actually.
This way both kind of developer are pleased.

detro commented on plugins/go-build/bin/goenv-install:
```diff
@@ -121,11 +129,23 @@ unset VERSION_NAME
-if grep -q -E "^[0-9]+\.[0-9]+(\s*)$" <<< ${DEFINITION}; then
-  echo "Adding patch version 0 to ${DEFINITION}"
-  DEFINITION="${DEFINITION}.0"
+# If latest is supplied, install the latest available (stable) version
+if [[ ${DEFINITION} == "latest" ]]; then
+  LATEST=$(latest_version "[0-9]\\.[0-9]+")
+  echo "Installing latest version ${LATEST}..."
+  DEFINITION=$LATEST
+# If unstable is supplied, install the latest available (including beta/rc) version
+elif [[ ${DEFINITION} == "unstable" ]]; then
+  LATEST=$(latest_includes_unstable_version "[0-9]\\.[0-9]+")
+  echo "Installing latest (including unstable) version ${LATEST}..."
+  DEFINITION=$LATEST
+# The latest patch version will be located, e.g if 1.11 is supplied they'll be changed to `1.11.x`.
+# NOTE: Try to capture semantic versions such as `1.11` which don't have a patch version and install latest patch.
```
Thank you for the inline doc.

detro reviewed (approved):
This looks really good.

EfChouTR commented:
@syndbg This looks ready to merge; tests all passed
